pangeo_forge_recipes/storage.py

```diff
@@ -98,7 +131,74 @@ def _full_path(self, path: str) -> str:
 class CacheFSSpecTarget(FlatFSSpecTarget):
     """Alias for FlatFSSpecTarget"""
 
-    pass
+    def cache_file(self, fname: str, **open_kwargs) -> None:
+        # check and see if the file already exists in the cache
+        logger.info(f"Caching file '{fname}'")
+        if self.exists(fname):
+            cached_size = self.size(fname)
+            remote_size = _get_url_size(fname)
+            if cached_size == remote_size:
+                # TODO: add checksumming here
+                logger.info(f"File '{fname}' is already cached")
+                return
+
+        input_opener = _fsspec_safe_open(fname, mode="rb", **open_kwargs)
+        target_opener = self.open(fname, mode="wb")
+        logger.info(f"Coping remote file '{fname}' to cache")
+        _copy_btw_filesystems(input_opener, target_opener)
+
+
+class MetadataTarget(FSSpecTarget):
+    """Target for storing metadata dictionaries as json."""
+
+    def __setitem__(self, key: str, value: dict) -> None:
+        mapper = self.get_mapper()
+        mapper[key] = json.dumps(value).encode("utf-8")
+
+    def __getitem__(self, key: str) -> dict:
+        return json.loads(self.get_mapper()[key])
+
+    def getitems(self, keys: Sequence[str]) -> dict:
+        mapper = self.get_mapper()
+        all_meta_raw = mapper.getitems(keys)
+        return {k: json.loads(raw_bytes) for k, raw_bytes in all_meta_raw.items()}
+
+
+@contextmanager
+def file_opener(
```

Comment:
Once merged, will this new method mean re-writing the **Step 5: Create storage targets** cell in the CMIP6 tutorial in some way? If so, I'll add this as a TODO for #135.

Reply:
This function (`file_opener`) is not something seen by the users. It's purely internal.

However, your comment makes me realize that there _is_ one user-facing change needed as a result of this. In step 5, instead of 
```python
meta_store = FSSpecTarget(fs_local, meta_dir.name)
```
we would need
```python
meta_store = MetadataTarget(fs_local, meta_dir.name)
```

Reply:
Great. Added this to item to #135.